orttraining/orttraining/core/framework/gradient_graph_builder.cc

```diff
@@ -127,16 +131,49 @@ Status GradientGraphBuilder::Build() {
   ORT_RETURN_IF_ERROR(opt_ret);
 
   GraphAugmenter::GraphDefs gradient_graph_defs;
-  // add "gradient of the loss" node, always 1.
-  if (loss_node_arg_name_ != "") {
+
+  // We skip adding loss gradient in case it's partitioned by pipeline.
+  if(COMMUNICATION_NODES.find(graph_->GetProducerNode(loss_node_arg_name_)->OpType()) == COMMUNICATION_NODES.end()) {
+    // We need to feed the loss output from forward graph and the gradient of loss initializer
+    // to the PassThrough op so that we manually enforce a dependency between backward pass
+    // and forward pass. Otherwise there will be cases where gradient ops don't rely on outputs
+    // from forward ops, then the backward pass might be scheduled to run before forward.
+    
+    // add "gradient of the loss" node, always 1.
     ONNX_NAMESPACE::TensorProto tensor_proto;
     tensor_proto.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_FLOAT);
     tensor_proto.add_float_data(1.f);
-    tensor_proto.set_name(GradientBuilderBase::GradientName(loss_node_arg_name_));
+    const std::string& loss_grad_name = GradientBuilderBase::GradientName(loss_node_arg_name_);
+    const std::string& loss_grad_initializer_name = loss_grad_name + "_initializer";
+    tensor_proto.set_name(loss_grad_initializer_name);
 
     gradient_graph_defs.AddInitializers({tensor_proto});
-  }
 
+    // Output of PassThrough is 'loss_name' + '_grad' which will be
+    // connected to first node in backward graph when we traverse
+    // through visited nodes later.
+    ArgDef loss_grad_initializer_arg = 
+        ArgDef(loss_grad_initializer_name,
+               gradient_graph_defs.CreateTypeProto({},
```

Comment:
loss_grad is not necessary a scalar, right?

Reply:
I think it should be. the loss output is a scalar so should its gradient?